adapters/rack/lib/opentelemetry/adapters/rack/middlewares/tracer_middleware.rb

```diff
@@ -0,0 +1,262 @@
+# frozen_string_literal: true
+
+# Copyright 2020 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require_relative '../util/queue_time'
+
+module OpenTelemetry
+  module Adapters
+    module Rack
+      module Middlewares
+        # TracerMiddleware propagates context and instruments Rack requests
+        # by way of its middleware system
+        #
+        # Notable implementation differences from dd-trace-rb:
+        # * missing: 'span.resource', which is set to span.name
+        # * missing: config[:distributed_tracing]
+        # * missing: span.set_error() -- spec level change
+        class TracerMiddleware # rubocop:disable Metrics/ClassLength
+          class << self
+            def allowed_rack_request_headers
+              @allowed_rack_request_headers ||= {}.tap do |result|
+                allowed_request_header_names.each do |header|
+                  result["HTTP_#{header.to_s.upcase.gsub(/[-\s]/, '_')}"] = build_attribute_name('http.request.headers.', header)
+                end
+              end
+            end
+
+            def allowed_request_header_names
+              @allowed_request_header_names ||= Array(config[:allowed_request_headers])
+            end
+
+            def allowed_response_headers
+              @allowed_response_headers ||= {}.tap do |result|
+                allowed_response_header_names.each do |header|
+                  result[header] = build_attribute_name('http.response.headers.', header)
+                  result[header.to_s.upcase] = build_attribute_name('http.response.headers.', header)
+                end
+              end
+            end
+
+            def allowed_response_header_names
+              @allowed_response_header_names ||= Array(config[:allowed_response_headers])
+            end
+
+            def build_attribute_name(prefix, suffix)
+              prefix + suffix.to_s.downcase.gsub(/[-\s]/, '_')
+            end
+
+            def config
+              Rack::Adapter.instance.config
+            end
+
+            private
+
+            def clear_cached_config
+              @allowed_rack_request_headers = nil
+              @allowed_request_header_names = nil
+
+              @allowed_response_headers = nil
+              @allowed_response_header_names = nil
+            end
+          end
+
+          EMPTY_HASH = {}.freeze
+
+          def initialize(app)
+            @app = app
+          end
+
+          def call(env) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
+            original_env = env.dup
+
+            parent_context = OpenTelemetry.tracer_factory.http_text_format.extract(env)
```

Comment:
Now that context prop has merged, this method is going to look something like this:

```ruby
def call(env)
  extracted_context = OpenTelemetry.propagation.extract(env)

  # restore extracted context in this process
  OpenTelemetry::Context.with_current(extracted_context) do

    frontend_context = create_frontend_span(env) if queue_time_present?

    request_span = tracer.start_span('request span', parent_context: frontend_context || extracted_context)

    tracer.with_span(request_span) do |span|
      begin
        @app.call(env)
      rescue StandardError => e
        span.set_error(e)
        raise e
      ensure
        add_attributes(request_span, ...)
        request_span.finish
        finish_frontend_span(frontend_context) if frontend_context
      end
    end
  end
end

# return 
def create_fontend_span(env)
  span = Tracer.start_span(...)
  extracted_context.with_value(current_span_key, span)
end

def finish_frontend_span(frontend_context)
  span = frontend_context[current_span_key]
  span.finish
end

def current_span_key
  OpenTelemetry::Trace::Propagation.current_span_key,
end
```

Reply:
I tried to get as close to these suggestions as possible, but there were a couple of hurdles.